### PR TITLE
Allow feature gates to be added to `etcd-druid` and vendor `etcd-druid@v0.19.1`

### PR DESCRIPTION
**Pull request by aaronfern**
**How to categorize this PR?**

/area backup
/kind api-change enhancement

**What this PR does / why we need it**:
With the release of `etcd-druid:v0.19.0`, feature gates are now supported in etcd-druid.
This PR updates the gardenlet config to allow featureGates to be specified in `GardenletConfiguration.EtcdConfig` and allows those feature gates to be passed on to the `etcd-druid` command

This PR also updated the vendored `etcd-druid` to `v0.19.0`

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
This PR should be merged only after #8299 has been merged

**Release note**:

```noteworthy operator
Gardenlet can now set feature gates for `etcd-druid`. They can be specified via the gardenlet configuration `GardenletConfiguration.EtcdConfig.FeatureGates`
```

### CODE REVIEW

File: pkg/gardenlet/apis/config/types.go [hunk 1]
```diff
@@ -457,6 +457,12 @@ type ETCDConfig struct {
 	BackupCompactionController *BackupCompactionController
 	// BackupLeaderElection contains configuration for the leader election for the etcd backup-restore sidecar.
 	BackupLeaderElection *ETCDBackupLeaderElection
+	// FeatureGates is a map of feature names to bools that enable or disable alpha/experimental
+	// features. This field modifies piecemeal the built-in default values from
+	// "github.com/gardener/etcd-druid/pkg/features/features.go".
+	// Default: nil
+	// +optional
```
Comment (timuthy) [inline pkg/gardenlet/apis/config/types.go]:
```suggestion
```

File: pkg/component/etcd/bootstrap.go [hunk 2]
```diff
@@ -401,6 +401,15 @@ func getDruidDeployCommands(etcdConfig *config.ETCDConfig) []string {
...
 	}
 
+	// Add feature gates to the etcd druid command
+	if etcdConfig.FeatureGates != nil {
+		var featureGateList string
+		for featuregate, featuregatevalue := range etcdConfig.FeatureGates {
+			featureGateList = featureGateList + featuregate + "=" + strconv.FormatBool(featuregatevalue) + ","
+		}
+		command = append(command, "--feature-gates="+featureGateList)
+	}
```
Comment (timuthy) [inline pkg/component/etcd/bootstrap.go]:
```suggestion
	if etcdConfig.FeatureGates != nil {
		command = append(command, kubernetesutils.FeatureGatesToCommandLineParameter(etcdConfig.FeatureGates))
	}
```

File: pkg/gardenlet/apis/config/v1alpha1/types.go [hunk 3]
```diff
@@ -555,6 +555,12 @@ type ETCDConfig struct {
 	// BackupLeaderElection contains configuration for the leader election for the etcd backup-restore sidecar.
 	// +optional
 	BackupLeaderElection *ETCDBackupLeaderElection `json:"backupLeaderElection,omitempty"`
+	// FeatureGates is a map of feature names to bools that enable or disable alpha/experimental
+	// features. This field modifies piecemeal the built-in default values from
+	// "github.com/gardener/etcd-druid/pkg/features/features.go".
+	// Default: nil
+	// +optional
+	FeatureGates map[string]bool `json:"featureGates,omitempty"`
```
Comment (rfranzke) [inline pkg/gardenlet/apis/config/v1alpha1/types.go:563]: please adapt gardenlet Helm chart as well as config example file in `example` directory

File: pkg/gardenlet/apis/config/v1alpha1/types.go [hunk 4]
```diff
@@ -555,6 +555,11 @@ type ETCDConfig struct {
 	// BackupLeaderElection contains configuration for the leader election for the etcd backup-restore sidecar.
 	// +optional
 	BackupLeaderElection *ETCDBackupLeaderElection `json:"backupLeaderElection,omitempty"`
+	// FeatureGates is a map of feature names to bools that enable or disable alpha/experimental
+	// features. This field modifies piecemeal the built-in default values from
+	// "github.com/gardener/etcd-druid/pkg/features/features.go".
+	// Default: nil
```
Comment (rfranzke) [inline pkg/gardenlet/apis/config/v1alpha1/types.go:561]:
```suggestion
	// Default: nil
	// +optional
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Review by rfranzke — COMMENTED** [review]
/lgtm

**Review by timuthy — COMMENTED** [review]
Thank you 🙂 
/lgtm
/approve

**Comment by timuthy** [comment]
The migration fails during restoration, when the system tries to copy the backup.

Log from copy job:

```
2023-08-04T07:35:42.721028967Z stderr F time="2023-08-04T07:35:42Z" level=info msg="etcd-backup-restore Version: v0.25.1"
2023-08-04T07:35:42.721062838Z stderr F time="2023-08-04T07:35:42Z" level=info msg="Git SHA: 7d97afa7"
2023-08-04T07:35:42.721067018Z stderr F time="2023-08-04T07:35:42Z" level=info msg="Go Version: go1.20.3"
2023-08-04T07:35:42.721069638Z stderr F time="2023-08-04T07:35:42Z" level=info msg="Go OS/Arch: linux/amd64"
2023-08-04T07:35:42.721329568Z stderr F time="2023-08-04T07:35:42Z" level=info msg="Temporary directory /var/etcd/data/tmp does not exist. Creating it..."
2023-08-04T07:35:42.721345728Z stderr F time="2023-08-04T07:35:42Z" level=fatal msg="Could not get source and destination snapstores: could not get destination snapstore: failed to create temporary directory /var/etcd/data/tmp: mkdir /var/etcd: permission denied"

```

@aaronfern could you please check?

**Comment by shreyas-s-rao** [comment]
/unhold since the error in e2e migration tests was seen only when the alpha feature `UseEtcdWrapper` is enabled (hence it's still an alpha feature), and *has no regression on the default behaviour.*

The e2e tests ended up with the feature gate set to true because this PR sets the example gardenlet config with `ETCDConfig.UseEtcdWrapper: true`. @timuthy @rfranzke  should we should keep it this way or change it to false for the e2e tests? My opinion is that it should be set to `false` by default in g/g code (deployment as well as e2e tests) since it's an `alpha` feature, and we would like to roll it out in a controlled manner before graduating it to `beta`. WDYT?

**Comment by ialidzhikov** [comment]
> The e2e tests ended up with the feature gate set to true because this PR sets the example gardenlet config with `ETCDConfig.UseEtcdWrapper: true`. @timuthy @rfranzke should we should keep it this way or change it to false for the e2e tests? My opinion is that it should be set to `false` by default in g/g code (deployment as well as e2e tests) since it's an `alpha` feature, and we would like to roll it out in a controlled manner before graduating it to `beta`. WDYT?

Agree, I would also vote to set it to `false` for now.

File: example/gardener-local/gardenlet/values.yaml [hunk 5]
```diff
@@ -49,6 +49,9 @@ config:
     # If disabled, a Shoot cluster in local setup won't be able to use the registry cache
     # (it will be still able to pull images but registry cache won't be used).
     ContainerdRegistryHostsDir: true
+  etcdConfig:
+    featureGates:
+      UseEtcdWrapper: true
```
Comment (shreyas-s-rao) [inline example/gardener-local/gardenlet/values.yaml]: Remove this since default behavior should be adhered to, ie, this feature gate will be set to `false`.

File: imagevector/images.yaml [hunk 6]
```diff
@@ -31,8 +31,8 @@ images:
         interacted with from other containers or other systems
 - name: etcd-druid
   sourceRepository: github.com/gardener/etcd-druid
-  repository: eu.gcr.io/gardener-project/gardener/etcd-druid
-  tag: "v0.19.0"
+  repository: aaronfernandes/druid #eu.gcr.io/gardener-project/gardener/etcd-druid
+  tag: "v0.19.0_copyjob_14" #"v0.19.0"
```
Comment (shreyas-s-rao) [inline imagevector/images.yaml]: Revert this change now that druid v0.19.1 is released.

File: pkg/gardenlet/apis/config/types.go [hunk 7]
```diff
@@ -457,6 +457,11 @@ type ETCDConfig struct {
 	BackupCompactionController *BackupCompactionController
 	// BackupLeaderElection contains configuration for the leader election for the etcd backup-restore sidecar.
 	BackupLeaderElection *ETCDBackupLeaderElection
+	// FeatureGates is a map of feature names to bools that enable or disable alpha/experimental
+	// features. This field modifies piecemeal the built-in default values from
+	// "github.com/gardener/etcd-druid/pkg/features/features.go".
+	// Default: nil
```
Comment (shreyas-s-rao) [inline pkg/gardenlet/apis/config/types.go:463]: Do we need to add `// +optional` here?

File: example/operator/10-componentconfig.yaml [hunk 8]
```diff
@@ -44,6 +44,8 @@ controllers:
         enableBackupCompaction: false
         eventsThreshold: 1000000
         activeDeadlineDuration: "3h"
+    # featureGates:
+    #   UseEtcdWrapper: true
```
Comment (shreyas-s-rao) [inline example/operator/10-componentconfig.yaml:48]: Either set this to `false` or remove it

**Review by shreyas-s-rao — CHANGES_REQUESTED** [review]
@aaronfern some changes required now after the druid v0.19.1. PTAL

File: go.mod [hunk 9]
```diff
@@ -10,7 +10,7 @@ require (
 	github.com/coreos/go-systemd/v22 v22.3.2
 	github.com/fluent/fluent-operator/v2 v2.2.0
 	github.com/gardener/dependency-watchdog v1.1.1
-	github.com/gardener/etcd-druid v0.18.4
+	github.com/gardener/etcd-druid v0.19.0
```
Comment (shreyas-s-rao) [inline go.mod]:
```suggestion
	github.com/gardener/etcd-druid v0.19.1
```

File: example/operator/10-componentconfig.yaml (hunk 8 above)
Comment (timuthy) [reply example/operator/10-componentconfig.yaml:48]: Actually, `true` should be ok since we usually follow the pattern `someKubernetesFeature: true` (see [here](https://github.com/gardener/gardener/blob/733e3d1b3242c5d3419ea34b38d610c6eb2e4f7c/example/90-shoot.yaml#L119))

File: pkg/gardenlet/apis/config/types.go (hunk 7 above)
Comment (ialidzhikov) [reply pkg/gardenlet/apis/config/types.go:463]: Nope, that's the internal type. We add it only to the external ones.

File: example/operator/10-componentconfig.yaml (hunk 8 above)
Comment (shreyas-s-rao) [reply example/operator/10-componentconfig.yaml:48]: Sure, then we can leave it as is, since default behavior is `false` anyway

File: pkg/gardenlet/apis/config/types.go (hunk 7 above)
Comment (shreyas-s-rao) [reply pkg/gardenlet/apis/config/types.go:463]: Ah got it

**Review by ialidzhikov — APPROVED** [review]
/lgtm
